PolicyOptimization/DistribPolicyGradients/ClientTrajectoryFinalizer.py

```diff
@@ -0,0 +1,148 @@
+    HEADER_INITIALIZATION="initialization"
+    HEADER_FLUSH = "flush_data"
+    HEADER_TRAJECTORY = "trajectory"
+
+    def __init__(self, loop_wait_period=0.1):
+        super().__init__("ClientTrajectoryFinalizer", loop_wait_period=0.1, process_all_updates=True)
+        self.handlers = {
+            ClientTrajectoryFinalizer.HEADER_INITIALIZATION: self._init,
+            ClientTrajectoryFinalizer.HEADER_FLUSH: self._flush,
+            ClientTrajectoryFinalizer.HEADER_TRAJECTORY: self._trajectory
+        }
+    
+    def init(self):
+        self.client = None
+        self.cfg = None
+        self.gamma = None
+        self.lmbda = None
+        self.value_estimator = None
+        self.reward_stats = None
+
+        self.trajectories_to_send = []
+        self.total_timesteps = 0
+        self.t0 = None
+        
+    def update(self, header, data):
+        handler = self.handlers.get(header, None)
+        if not handler:
+            print(f"WARNING: ClientTrajectoryFinalizer received unknown message header '{header}'")
+            return
+
+        handler(data)
+
+    def _init(self, cfg):
+        self.cfg = cfg
+        self.gamma = self.cfg["policy_optimizer"]["gamma"]
+        self.lmbda = self.cfg["policy_optimizer"]["gae_lambda"]
+
+        pf_cfg = {
+            "device": cfg.get("device", "cpu"),
+            "value_estimator": cfg["value_estimator"]
+        }
+
+        models = PolicyFactory.get_from_cfg(pf_cfg, env_space_shapes=cfg["env_space_shapes"])
+        self.value_estimator = models["value_estimator"]
+
+        if self.client is not None:
+            try:
+                self.client.disconnect()
+            except:
+                pass
+
+        self.client = RedisClient()
+        self.client.connect()
+
+        value_params = None
+        while value_params is None:
+            value_params = self.client.get_latest_value_params()
+
+        print(f"updated value estimator to version {self.client.current_value_epoch}")
+
+        self.value_estimator.set_trainable_flat(value_params)
+        self.reward_stats = self.client.get_reward_stats()
+
+        self.trajectories_to_send = []
+        self.total_timesteps = 0
+    
+    def _flush(self, rewards):
+        if self._server_is_running():
+            t1 = time.perf_counter()
+            encoded = cser.pack(self.trajectories_to_send)
+
+            self.client.push_data(RedisKeys.CLIENT_EXPERIENCE_KEY, encoded, encoded=True)
+
+            t2 = time.perf_counter()
+
+            # we use t1 here because we don't count the time involved in the
+            # send, as it's not blocking the game
+            seconds = t1 - self.t0
+            steps_per_second = float(self.total_timesteps) / seconds
+
+            if rewards:
+                self.client.push_data(RedisKeys.CLIENT_POLICY_REWARD_KEY, cser.pack(rewards), encoded=True)
+
+            print("packed and pushed {} trajectories containing {} timesteps in {:7.5f}s ({:.2f} sps)".format(

```

Comment:
Note that the first time here is just the time that it took to pack and push the trajectories and rewards. Let me know if this is too verbose, or if there's some other way that you'd like it to be called out